merlin/table/tensor_table.py

```diff
@@ -0,0 +1,181 @@
+#
+# Copyright (c) 2023, NVIDIA CORPORATION.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+from typing import Any, Dict
+
+from merlin.dag.utils import group_values_offsets
+from merlin.table.cupy_column import CupyColumn
+from merlin.table.numpy_column import NumpyColumn
+from merlin.table.tensor_column import TensorColumn, create_tensor_column
+from merlin.table.tensorflow_column import TensorflowColumn
+from merlin.table.torch_column import TorchColumn
+
+TensorDict = Dict[str, Any]
+ColumnDict = Dict[str, TensorColumn]
+
+
+class TensorTable:
+    """
+    A DataFrameLike wrapper around a dictionary of arrays or tensors
+    """
+
+    def __init__(self, columns: TensorDict = None):
+        grouped_columns = group_values_offsets(columns or {})
+        dict_cols = {}
+        for name, column in grouped_columns.items():
+            if isinstance(column, TensorColumn):
+                dict_cols[name] = column
+            elif isinstance(column, tuple):
+                dict_cols[name] = create_tensor_column(column[0], column[1])
+            else:
+                dict_cols[name] = create_tensor_column(column)
+
+        self._columns = dict_cols
+
+    def __iter__(self):
+        return iter(self._columns)
+
+    def __len__(self):
+        return len(self._columns)
+
+    def __getitem__(self, key):
+        if isinstance(key, list):
+            return TensorTable({k: self._columns[k] for k in key})
+        else:
+            return self._columns[key]
+
+    def __setitem__(self, key, value):
+        self._columns[key] = value
+
+    def __delitem__(self, key):
+        del self._columns[key]
+
+    @property
+    def device(self):
+        return list(self.values())[0].device
```

Comment:
it seems like it's currently possible to construct a TensorTable with columns of different types and potentially on different devices. If that's something we'd like to support then this device property may need to change. to return for example a dictionary of devices by column name?

Reply:
I think we should add validation to prevent that 👍🏻

Reply:
Adding a restriction on the same device seems reasonable. What about tensors of different types that are on the same device, should we restrict not only to the same device but also to the same column type?

Reply:
Yeah, I think so. I've added both validations in the latest commits